### ISSUE TITLE
Variable length generation

### ISSUE THREAD

**Issue by xtwigs** [jxiw/varlen_mamba#1]
Hi Junxiong,

I'm trying to work with variable length generation for Mamba-1 but not quite sure how to extract the ssm_state per sample in the batch.

Currently, when passing `position_indices = [0,1,..., n, 0, ..., n]` (effectively no padding)  the outputs are identical but the ssm_state only gets returned for the last sample in the packed sequence. 

```
from mamba_ssm.utils.generation import InferenceParams
from fla.layers.utils import get_unpad_data, index_first_axis, pad_input

bsz = 2
seqlen = 16
d_model = 32

inf1 = InferenceParams(max_batch_size=bsz, max_seqlen=32) # max_seqlen is ignored
inf2 = InferenceParams(max_batch_size=bsz, max_seqlen=32)

x = torch.randn(bsz, seqlen, d_model)
att_mask = torch.ones(bsz, seqlen)

if att_mask is not None:
    indices, cu_seqlens, _ = get_unpad_data(att_mask[:, -seqlen:])
    x1 = index_first_axis(rearrange(x, "b s ... -> (b s) ..."), indices)

position_indices = torch.cat([torch.arange(l, dtype=torch.int32, device=x.device) for l in cu_seqlens[1:] - cu_seqlens[:-1]], dim=0)
x1, x, position_indices= x1.cuda(), x.cuda(), position_indices.cuda()

x1, position_indices = x1.unsqueeze(0), position_indices.unsqueeze(0)

out1 = m1.forward(x1, position_indices=position_indices, inference_params=inf1)
out2 = m1.forward(x,  inference_params=inf2)

y1 = pad_input(out1[0], indices, bsz, seqlen)

out1.shape, out2.shape, (y1 - out2).abs().mean()
```

returns `(torch.Size([1, 32, 32]),
 torch.Size([2, 16, 32]),
 tensor(1.8578e-09, device='cuda:0', grad_fn=<MeanBackward0>))`

But inspecting the ssm_state only the last sample gets returned:

```
inf1.key_value_memory_dict[0][1].shape, inf2.key_value_memory_dict[0][1].shape

# (torch.Size([1, 64, 16]), torch.Size([2, 64, 16]))
```

I found the same problem in this PR from @zigzagcai  (https://github.com/state-spaces/mamba/pull/244#issuecomment-2874082895). How did you fix this issue?

**Comment by jxiw** [comment]
you are absolutely correct. it only returns the last sample, so it can only use in the training. you can refer to the solution in vllm https://github.com/vllm-project/vllm/blob/main/csrc/mamba/mamba_ssm/selective_scan_fwd.cu#L258. currently, i am busy with other staff. i will try to fix this once i get some time.

**Comment by xtwigs** [comment]
Thank you for the quick response!

I was finding it odd that no one was considering generation. Looking forward to seeing more of your great work.

**Comment by jxiw** [comment]
yeah. i agree it is very important to support generations using variable length kernel + cuda graph. i am working on that, but get distracted for some other stuff. but i agree with your vision.

**Comment by jxiw** [comment]
Hi @xtwigs, i find sometime to work on that. this feature is support now. example usage.

```

batch_size = int((position_ids == 0).sum())

ssm_state = torch.zeros(batch_size, self.d_inner, self.d_state, device=x.device, dtype=x.dtype)

y = selective_scan_fn(
                x,
                dt,
                A,
                B,
                C,
                self.D.float(),
                z=z,
                delta_bias=self.dt_proj.bias.float(),
                delta_softplus=True,
                return_last_state=True,
                position_indices=position_ids,
                ssm_state=ssm_state, 
                cu_seqlens=cu_seqlens,
            )

ssm_state = rearrange(ssm_state, "b (h d) n -> b h d n", h=n_groups) // in case, you have n_group.
```

i will add some tests on here https://github.com/jxiw/M1 later. thank you.
